loadbalancer/algorithm.go

```diff
@@ -89,19 +96,70 @@ func (*consistentHash) Apply(ctx *routing.LBContext) routing.LBEndpoint {
 	key := net.RemoteHost(ctx.Request).String()
 	if _, err := h.Write([]byte(key)); err != nil {
 		log.Errorf("Failed to write '%s' into hash: %v", key, err)
-		return ctx.Route.LBEndpoints[rand.Intn(len(ctx.Route.LBEndpoints))]
+		return ctx.Route.LBEndpoints.At(rand.Intn(ctx.Route.LBEndpoints.Length()))
 	}
 	sum = h.Sum32()
-	choice := int(sum) % len(ctx.Route.LBEndpoints)
+	choice := int(sum) % ctx.Route.LBEndpoints.Length()
 	if choice < 0 {
-		choice = len(ctx.Route.LBEndpoints) + choice
+		choice = ctx.Route.LBEndpoints.Length() + choice
+	}
+	return ctx.Route.LBEndpoints.At(choice)
+}
+
+type powerOfChoices struct{
+	rand *rand.Rand
+	numberOfChoices int
+	mutex sync.Mutex
+}
+
+// Selects N random backends and picks the one with less outstanding requests.
+func newPowerOfChoices(endpoints []string) routing.LBAlgorithm {
+	t := time.Now().UnixNano()
+	return &powerOfChoices{
+		rand: rand.New(rand.NewSource(t)),
+		numberOfChoices: 2, // TODO: make this the value part of skipper configuration.
+	}
+}
+func (a *powerOfChoices) GetRandomChoice(length int) int {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	return a.rand.Intn(length)
+}
+
+func contains(s []int, e int) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
 	}
-	return ctx.Route.LBEndpoints[choice]
+	return false
+}
+
+func (a *powerOfChoices) Apply(ctx *routing.LBContext) routing.LBEndpoint {
+	chosen := make([]int, 0, a.numberOfChoices)
+	for i := 0 ; i < a.numberOfChoices ; i++  {
+		candidate := a.GetRandomChoice(ctx.Route.LBEndpoints.Length())
+		// Pick a different endpoint if was already selected
```

Comment:
The paper [The Power of Two Choices in Randomized Load Balancing](https://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf) explores the algorithm based on the assumption that _each ball is placed sequentially into the least full of d bins chosen independently and uniformly at random_ (ball is the request and bin is the route in case of skipper). Here the choice of bins (routes) is not independent.

Moreover, consider the corner case n=1 (only one route available) and d=2 (hardcoded two draws): in this case `for contains(chosen, candidate)` becomes infinite loop on the second draw.

With the above said, proposed implementation could be simplified to draw routes independently and would require O(1) memory and O(d) cpu instead of O(d) memory and O(d^d) cpu needed to remember previous choices.

It also seems to be a good idea to extend `Random` algorithm with a draw count parameter like [it was done in HAProxy](https://github.com/haproxy/haproxy/commit/21c741a665f) instead of introducing new algorithm.

Reply:
Very good observation. I missed the `independently and uniformly` part of the algorithm. I am fixing the issue and trying to unify it to the random implementation.

Reply:
I fixed the algo however unifying random and power of N choices could be address in a separate PR as it requires to change the initialisation mechanism for LB algorithms. Currently there is no way to parametrise algorithms other than with endpoints.